subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/beans/BeanPropertyWriter.kt

```diff
@@ -69,9 +73,25 @@ class BeanPropertyWriter(
 }
 
 
+private
+fun WriteContext.reportFieldProblems(field: Field, fieldName: String, fieldType: Class<*>, fieldValue: Any?) {
+    if (!field.isAnnotationPresent(DisableInstantExecutionFieldTypeCheck::class.java)) {
+        withPropertyTrace(PropertyKind.Field, fieldName) {
+            unsupportedFieldDeclaredTypes
```

Comment:
I would consider doing the inspection in `relevantStateOf()`, so we do the check once per type, rather than once per instance of that type.

Reply:
We want the trace to be reported on each occurrence of the problem so it's easier to find the offenders. So, the check has to be done on each instance of the type.

What we could do is cache the result of the check per type and still emit the problem on each instance of the type. But that would require keeping already seen bean types and the result of the check. Do you think it worth it?

Reply:
I implemented this in https://github.com/gradle/gradle/pull/12760/commits/8268481037fcf8e4587229a0c9eb8a40f49d0a5e